bindings/src/main/java/io/pravega/storage/filesystem/FileSystemStorage.java

```diff
@@ -458,5 +466,225 @@ private Void doDelete(SegmentHandle handle) throws IOException {
+
+            // Check temp file already exists.
+            String tmpSegmentName = getTempSegmentName(segmentName);
+            if (super.doExists(tmpSegmentName)) {
+                if (baseExists) {
+                    // We have both a temp file and a segment file. This is most likely the result of an incomplete replace,
+                    // however we still have the original segment file around. It is safe to delete the temp file now.
+                    log.debug("Incomplete replace operation detected for '{}'. Deleting temp file before new replace attempt.", segmentName);
+                    super.doDelete(super.doOpenWrite(tmpSegmentName));
+                } else {
+                    // Temp file exists, but the segment file does not. This may be the result of an incomplete replace,
+                    // in which case we need to finalize that one to prevent deleting (what could be) the only persisted
+                    // copy of our data.
+                    log.debug("Incomplete replace operation detected for '{}'. Finalizing before new replace attempt.", segmentName);
+                    finalizeRename(tmpSegmentName, segmentName);
+                }
+            } else if (!baseExists) {
+                throw new StreamSegmentNotExistsException(segmentName);
+            }
+
+            // Write given contents to temp file.
+            val tmpHandle = super.doCreate(tmpSegmentName);
+            try {
+                super.doWrite(tmpHandle, 0, contents.getReader(), contents.getLength());
+                if (shouldReseal) {
+                    super.doSeal(tmpHandle);
+                }
+            } catch (Exception ex) {
+                log.warn("Unable to write to temporary file when attempting to replace '{}'. Original file has not been touched. Cleaning up.",
+                        segmentName, ex);
+                super.doDelete(tmpHandle);
+                throw ex;
+            }
+
+            // Sanity check #1 (only executes in tests).
+            assert super.doGetStreamSegmentInfo(tmpSegmentName).getLength() == contents.getLength();
+
+            // Rename file. After this is done, the replace operation is complete.
+            finalizeRename(tmpSegmentName, segmentName);
+
+            // Sanity check #2 (only executes in tests).
+            assert super.doGetStreamSegmentInfo(segmentName).getLength() == contents.getLength();
+            return null;
+        }
+
+        @SneakyThrows(StreamSegmentReplaceException.class)
+        private void finalizeRename(String fromSegmentName, String toSegmentName) {
+            val fromFile = getPath(fromSegmentName).toFile().getAbsoluteFile();
+            val toFile = getPath(toSegmentName).toFile().getAbsoluteFile();
+            boolean renamed = fromFile.renameTo(toFile);
+            if (!renamed) {
+                // On Windows, renameTo does not replace.
```

Comment:
When we say `On Windows...`, is it referring to the Windows implementation of the JVM or a Windows file system (FAT, NTFS)? I ask because it is the LTS implementation that needs to implement rename, right?

Also, is the behavior reflected here documented anywhere we can point to or was this validated manually?

Reply:
The Javadoc for `File.renameTo` states that this operation may not be atomic or may not succeed if a file with the destination name already exists, so that is enough to warrant the `if` block and the subsequent 2-step rename.
 
With respect to Windows, I am unaware of the inner details on what the requirements are. For this code, I got inspired from the AtomicFileOutputStream (both in the [HDFS repository](https://github.com/apache/hadoop/blob/trunk/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/util/AtomicFileOutputStream.java) and [ZooKeeper repository](https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/common/AtomicFileOutputStream.java) - they look nearly identical save for a few diffs).

Reply:
Added some clarifying comment on this.